### PR TITLE
fix(components): [menu] menus that are not in the expanded path do not open

### LINKED ISSUE #10431: [Component] [menu, sub-menu] In el menu, if the submenu data under the primary menu is obtained dynamically, loading the submenu data for the first time will cause the selected but closed menu page to be expanded
Bug Type: **`Component`**

## Environment

- Vue Version: `3.2.41`
- Element Plus Version: `2.2.19`
- Browser / OS: `Chrome 107.0.5304.88`
- Build Tool: `Webpack`

### Related Component

- `el-menu`
- `el-sub-menu`

### Reproduction Link

[Element Plus Playground](https://element-plus.run/#eyJBcHAudnVlIjoiPHRlbXBsYXRlPlxuICA8ZGl2IGNsYXNzPVwiY29tbW9uLWxheW91dFwiPlxuICAgIDxlbC1tZW51PlxuICAgICAgPGVsLXN1Yi1tZW51IHYtZm9yPVwiKG1lbnUsIGluZGV4KSBvZiBtZW51RGF0YVwiIDppbmRleD1cImAke2luZGV4fWBcIiBAY2xpY2s9XCJoYW5kbGVPcGVuU3ViTWVudShpbmRleClcIj5cbiAgICAgICAgPHRlbXBsYXRlICN0aXRsZT57e21lbnUudGl0bGV9fTwvdGVtcGxhdGU+XG4gICAgICAgIDxlbC1zdWItbWVudSA6aW5kZXg9XCJgJHtpbmRleH0tMGBcIiB2LWlmPVwibmF2RGF0YVttZW51LnRpdGxlXVwiPlxuICAgICAgICAgIDx0ZW1wbGF0ZSAjdGl0bGU+e3tuYXZEYXRhW21lbnUudGl0bGVdLnN1YlRpdGxlfX08L3RlbXBsYXRlPlxuICAgICAgICAgIDxlbC1tZW51LWl0ZW0gOmluZGV4PVwiYCR7aW5kZXh9LTAtMGBcIj5cbiAgICAgICAgICAgIDx0ZW1wbGF0ZSAjdGl0bGU+e3tuYXZEYXRhW21lbnUudGl0bGVdLml0ZW1UaXRsZX19PC90ZW1wbGF0ZT5cbiAgICAgICAgICA8L2VsLW1lbnUtaXRlbT5cbiAgICAgICAgPC9lbC1zdWItbWVudT5cbiAgICAgIDwvZWwtc3ViLW1lbnU+XG4gICAgPC9lbC1tZW51PlxuICA8L2Rpdj5cbjwvdGVtcGxhdGU+XG5cbjxzY3JpcHQgc2V0dXA+XG5pbXBvcnQge3JlYWN0aXZlfSBmcm9tIFwidnVlXCI7XG5cbmNvbnN0IG1lbnVEYXRhID0gcmVhY3RpdmUoW3tcbiAgdGl0bGU6ICdtZW51J1xufV0pXG5jb25zdCBuYXZEYXRhID0gcmVhY3RpdmUoe30pXG5mb3IgKGxldCBpPTA7IGk8NzsgaSsrKSB7XG4gIG1lbnVEYXRhLnB1c2goey4uLm1lbnVEYXRhWzBdLCB0aXRsZTogYG1lbnUke2kgKyAxfWB9KVxufVxuY29uc3QgaGFuZGxlT3BlblN1Yk1lbnUgPSBpbmRleCA9PiB7XG4gIGNvbnN0IG1lbnUgPSBtZW51RGF0YVtpbmRleF1cbiAgaWYgKG5hdkRhdGFbbWVudS50aXRsZV0pIHJldHVyblxuICBuYXZEYXRhW21lbnUudGl0bGVdID0ge1xuICAgIHN1YlRpdGxlOiBgc3ViVGl0bGUke2luZGV4fWAsXG4gICAgaXRlbVRpdGxlOiBgbWVudUl0ZW0ke2luZGV4fWBcbiAgfVxufVxuPC9zY3JpcHQ+XG5cbjxzdHlsZSBzY29wZWQ+XG48L3N0eWxlPlxuXG4iLCJpbXBvcnRfbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7fVxufSIsIl9vIjp7fX0=)

### Steps to reproduce

First expand a menu page randomly, select a menu item, and then close the menu page to the unexpanded state. Then click another menu page which haven't loaded submenu data yet, and you will find that the menu page that was selected but closed has also been expanded

### What is Expected?

The menu page that was originally selected and closed remains closed

### What is actually happening?

The menu page that was originally selected and closed is inexplicably expanded

### Additional comments

_(empty)_

### PR DESCRIPTION
**Pull request by iDestin**
fix #10431 and unify the hooks into the hooks directory

closed #10431

Please make sure these boxes are checked before submitting your PR, thank you!

- [x] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [x] Make sure you are merging your commits to `dev` branch.
- [x] Add some descriptions and refer to relative issues for your PR.

### CODE REVIEW

File: docs/.vitepress/vitepress/styles/css-vars.scss [hunk 1]
```diff
@@ -68,7 +68,7 @@
   --code-text-color: var(--text-color);
   --code-font-family: var(--font-family-mono);
   --code-tooltip-bg-color: var(--code-bg-color);
-  --code-tooltip-color: #0C61C9;
+  --code-tooltip-color: #0c61c9;
```
Comment (holazz) [inline docs/.vitepress/vitepress/styles/css-vars.scss:71]: Please revert this change.

File: docs/en-US/component/message-box.md [hunk 2]
```diff
@@ -191,4 +191,4 @@ The corresponding methods are: `ElMessageBox`, `ElMessageBox.alert`, `ElMessageB
...
-| append-to              | set the root element for the message box                                                                                                         | string \| HTMLElement                                                                                                                                                                                                                         | —                                | —                                        |
+| append-to                    | set the root element for the message box                                                                                                 | string \| HTMLElement                                                                                                                                                                                                           | —                                | —                                                |
```
Comment (holazz) [inline docs/en-US/component/message-box.md:194]: Same here.

**Review by holazz — COMMENTED** [review]
Hi @iDestin , I think we shouldn't introduce changes that are unrelated to this fix. For extract hooks, you can do it in a separate PR.

File: docs/.vitepress/vitepress/styles/css-vars.scss (hunk 1 above)
Comment (iDestin) [reply docs/.vitepress/vitepress/styles/css-vars.scss:71]: lol... just run pnpm format😂

**Comment by iDestin** [comment]
ok I'll handle it later

**Comment by iDestin** [comment]
PTAL 🙇 @holazz

**Comment by holazz** [comment]
Seems like there is still a bug in `unique-opend` mode. [Demo](https://element-plus.run/#eyJBcHAudnVlIjoiPHRlbXBsYXRlPlxuICA8ZGl2IGNsYXNzPVwiY29tbW9uLWxheW91dFwiPlxuICAgIDxlbC1tZW51IHVuaXF1ZS1vcGVuZWQgZGVmYXVsdC1hY3RpdmU9XCIwLTAtMFwiPlxuICAgICAgPGVsLXN1Yi1tZW51IHYtZm9yPVwiKG1lbnUsIGluZGV4KSBvZiBtZW51RGF0YVwiIDppbmRleD1cImAke2luZGV4fWBcIiBAY2xpY2s9XCJoYW5kbGVPcGVuU3ViTWVudShpbmRleClcIj5cbiAgICAgICAgPHRlbXBsYXRlICN0aXRsZT57e21lbnUudGl0bGV9fTwvdGVtcGxhdGU+XG4gICAgICAgIDxlbC1zdWItbWVudSA6aW5kZXg9XCJgJHtpbmRleH0tMGBcIiB2LWlmPVwibmF2RGF0YVttZW51LnRpdGxlXVwiPlxuICAgICAgICAgIDx0ZW1wbGF0ZSAjdGl0bGU+e3tuYXZEYXRhW21lbnUudGl0bGVdLnN1YlRpdGxlfX08L3RlbXBsYXRlPlxuICAgICAgICAgIDxlbC1tZW51LWl0ZW0gOmluZGV4PVwiYCR7aW5kZXh9LTAtMGBcIj5cbiAgICAgICAgICAgIDx0ZW1wbGF0ZSAjdGl0bGU+e3tuYXZEYXRhW21lbnUudGl0bGVdLml0ZW1UaXRsZX19PC90ZW1wbGF0ZT5cbiAgICAgICAgICA8L2VsLW1lbnUtaXRlbT5cbiAgICAgICAgPC9lbC1zdWItbWVudT5cbiAgICAgIDwvZWwtc3ViLW1lbnU+XG4gICAgPC9lbC1tZW51PlxuICA8L2Rpdj5cbjwvdGVtcGxhdGU+XG5cbjxzY3JpcHQgc2V0dXA+XG5pbXBvcnQge3JlYWN0aXZlfSBmcm9tIFwidnVlXCI7XG5cbmNvbnN0IG1lbnVEYXRhID0gcmVhY3RpdmUoW3tcbiAgdGl0bGU6ICdtZW51J1xufV0pXG5jb25zdCBuYXZEYXRhID0gcmVhY3RpdmUoe30pXG5mb3IgKGxldCBpPTA7IGk8NzsgaSsrKSB7XG4gIG1lbnVEYXRhLnB1c2goey4uLm1lbnVEYXRhWzBdLCB0aXRsZTogYG1lbnUke2kgKyAxfWB9KVxufVxuY29uc3QgaGFuZGxlT3BlblN1Yk1lbnUgPSBpbmRleCA9PiB7XG4gIGNvbnN0IG1lbnUgPSBtZW51RGF0YVtpbmRleF1cbiAgaWYgKG5hdkRhdGFbbWVudS50aXRsZV0pIHJldHVyblxuICBuYXZEYXRhW21lbnUudGl0bGVdID0ge1xuICAgIHN1YlRpdGxlOiBgc3ViVGl0bGUke2luZGV4fWAsXG4gICAgaXRlbVRpdGxlOiBgbWVudUl0ZW0ke2luZGV4fWBcbiAgfVxufVxuPC9zY3JpcHQ+XG5cbjxzdHlsZSBzY29wZWQ+XG48L3N0eWxlPlxuXG4iLCJQbGF5Z3JvdW5kTWFpbi52dWUiOiI8c2NyaXB0IHNldHVwPlxuaW1wb3J0IEFwcCBmcm9tICcuL0FwcC52dWUnXG5pbXBvcnQgeyBzZXR1cEVsZW1lbnRQbHVzIH0gZnJvbSAnLi9lbGVtZW50LXBsdXMuanMnXG5zZXR1cEVsZW1lbnRQbHVzKClcbjwvc2NyaXB0PlxuXG48dGVtcGxhdGU+XG4gIDxBcHAgLz5cbjwvdGVtcGxhdGU+XG4iLCJpbXBvcnRfbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7XG4gICAgXCJlbGVtZW50LXBsdXNcIjogXCJodHRwczovL3ByZXZpZXctMTA0ODAtZWxlbWVudC1wbHVzLnN1cmdlLnNoL2J1bmRsZS9pbmRleC5mdWxsLm1pbi5tanNcIixcbiAgICBcImVsZW1lbnQtcGx1cy9cIjogXCJ1bnN1cHBvcnRlZFwiXG4gIH1cbn0iLCJpbXBvcnQtbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7XG4gICAgXCJ2dWVcIjogXCJodHRwczovL2Zhc3RseS5qc2RlbGl2ci5uZXQvbnBtL0B2dWUvcnVudGltZS1kb21AbGF0ZXN0L2Rpc3QvcnVudGltZS1kb20uZXNtLWJyb3dzZXIuanNcIixcbiAgICBcIkB2dWUvc2hhcmVkXCI6IFwiaHR0cHM6Ly9mYXN0bHkuanNkZWxpdnIubmV0L25wbS9AdnVlL3NoYXJlZEBsYXRlc3QvZGlzdC9zaGFyZWQuZXNtLWJ1bmRsZXIuanNcIixcbiAgICBcImVsZW1lbnQtcGx1c1wiOiBcImh0dHBzOi8vcHJldmlldy0xMDQ4MC1lbGVtZW50LXBsdXMuc3VyZ2Uuc2gvYnVuZGxlL2luZGV4LmZ1bGwubWluLm1qc1wiLFxuICAgIFwiZWxlbWVudC1wbHVzL1wiOiBcInVuc3VwcG9ydGVkXCIsXG4gICAgXCJAZWxlbWVudC1wbHVzL2ljb25zLXZ1ZVwiOiBcImh0dHBzOi8vZmFzdGx5LmpzZGVsaXZyLm5ldC9ucG0vQGVsZW1lbnQtcGx1cy9pY29ucy12dWVAMi9kaXN0L2luZGV4Lm1pbi5qc1wiXG4gIH0sXG4gIFwic2NvcGVzXCI6IHt9XG59IiwiZWxlbWVudC1wbHVzLmpzIjoiaW1wb3J0IHsgZ2V0Q3VycmVudEluc3RhbmNlIH0gZnJvbSAndnVlJ1xuaW1wb3J0IEVsZW1lbnRQbHVzIGZyb20gJ2VsZW1lbnQtcGx1cydcblxubGV0IGluc3RhbGxlZCA9IGZhbHNlXG5hd2FpdCBsb2FkU3R5bGUoKVxuXG5leHBvcnQgZnVuY3Rpb24gc2V0dXBFbGVtZW50UGx1cygpIHtcbiAgaWYgKGluc3RhbGxlZCkgcmV0dXJuXG4gIGNvbnN0IGluc3RhbmNlID0gZ2V0Q3VycmVudEluc3RhbmNlKClcbiAgaW5zdGFuY2UuYXBwQ29udGV4dC5hcHAudXNlKEVsZW1lbnRQbHVzKVxuICBpbnN0YWxsZWQgPSB0cnVlXG59XG5cbmV4cG9ydCBmdW5jdGlvbiBsb2FkU3R5bGUoKSB7XG4gIHJldHVybiBuZXcgUHJvbWlzZSgocmVzb2x2ZSwgcmVqZWN0KSA9PiB7XG4gICAgY29uc3QgbGluayA9IGRvY3VtZW50LmNyZWF0ZUVsZW1lbnQoJ2xpbmsnKVxuICAgIGxpbmsucmVsID0gJ3N0eWxlc2hlZXQnXG4gICAgbGluay5ocmVmID0gJ2h0dHBzOi8vcHJldmlldy0xMDQ4MC1lbGVtZW50LXBsdXMuc3VyZ2Uuc2gvYnVuZGxlL2luZGV4LmNzcydcbiAgICBsaW5rLmFkZEV2ZW50TGlzdGVuZXIoJ2xvYWQnLCByZXNvbHZlKVxuICAgIGxpbmsuYWRkRXZlbnRMaXN0ZW5lcignZXJyb3InLCByZWplY3QpXG4gICAgZG9jdW1lbnQuYm9keS5hcHBlbmQobGluaylcbiAgfSlcbn0iLCJfbyI6eyJzaG93SGlkZGVuIjp0cnVlLCJzdHlsZVNvdXJjZSI6Imh0dHBzOi8vcHJldmlldy0xMDQ4MC1lbGVtZW50LXBsdXMuc3VyZ2Uuc2gvYnVuZGxlL2luZGV4LmNzcyJ9fQ==)

https://user-images.githubusercontent.com/23100055/200719834-41882666-559d-4567-94c2-d6b950f23640.mov

**Comment by holazz** [comment]
> not caused by this change, should I open a separate issue?

I guess they are both issues caused by dynamically generating menu items, so I think it should be fixed in one PR.

**Comment by iDestin** [comment]
Do you have time? Please take a look @holazz @chenxch

**Comment by holazz** [comment]
@iDestin Sorry, I've been busy lately, I'll review it later. In addition, if there is no conflict with the dev branch, you don't need to merge the dev branch into current.

**Comment by iDestin** [comment]
> @iDestin Sorry, I've been busy lately, I'll review it later. In addition, if there is no conflict with the dev branch, you don't need to merge the dev branch into current.

oh sorry! i dont know this, thk

File: packages/components/menu/src/menu.ts [hunk 3]
```diff
@@ -116,7 +116,7 @@ export default defineComponent({
     const activeIndex = ref<MenuProvider['activeIndex']>(props.defaultActive)
     const items = ref<MenuProvider['items']>({})
     const subMenus = ref<MenuProvider['subMenus']>({})
-
+    const activeItemFirstOpend = ref<boolean>(true)
```
Comment (holazz) [inline packages/components/menu/src/menu.ts:119]: It seems that there is no need to use `ref`
```suggestion
    let activeItemFirstOpend = true
```

File: packages/components/menu/src/menu.ts [hunk 4]
```diff
@@ -128,16 +128,37 @@ export default defineComponent({
...
       const indexPath = activeItem.indexPath
 
       // 展开该菜单项的路径上所有子菜单
       // expand all subMenus of the menu item
-      indexPath.forEach((index) => {
-        const subMenu = subMenus.value[index]
-        subMenu && openMenu(index, subMenu.indexPath)
-      })
+      const openAllActiveItem = () => {
+        if (activeItemFirstOpend.value) activeItemFirstOpend.value = false
```
Comment (holazz) [inline packages/components/menu/src/menu.ts:144]: The judgment condition here also seems unnecessary.
```suggestion
        activeItemFirstOpend = false
```

**Review by holazz — COMMENTED** [review]
Hi, @iDestin I left a few comments for your consideration. Also, would you mind changing the test case to be dynamically generated for the menu item?

**Comment by iDestin** [comment]
> Hi, @iDestin I left a few comments for your consideration. Also, would you mind changing the test case to be dynamically generated for the menu item?

done 👀

**Review by holazz — APPROVED** [review]

**Review by tolking — APPROVED** [review]
